src/libraries/Microsoft.Extensions.Options/src/OptionsCache.cs

```diff
@@ -45,6 +45,29 @@ public virtual TOptions GetOrAdd(string? name, Func<TOptions> createOptions!!)
             return value.Value;
         }
 
+        internal TOptions GetOrAdd<TArg>(string? name, Func<string, TArg, TOptions> createOptions, TArg factoryArgument)
+        {
+            // For compatibility, fall back to public GetOrAdd() if we're in a derived class.
+            if (GetType() != typeof(OptionsCache<TOptions>))
+            {
+                return GetOrAdd(name, () => createOptions(name ?? Options.DefaultName, factoryArgument));
+            }
+
+            name ??= Options.DefaultName;
+            Lazy<TOptions> value;
+
+#if NET || NETSTANDARD2_1
+            value = _cache.GetOrAdd(name, static (name, arg) => new Lazy<TOptions>(arg.createOptions(name, arg.factoryArgument)), (createOptions, factoryArgument));
```

Comment:
This would be less awkward if we weren't using `Lazy`. From what I can tell the benefit we're getting from that is ensuring that we never create an options instance that doesn't end up getting used. Is that worth it?